### ISSUE TITLE
List & Dict can't accept `array.dtype` as argument.

### ISSUE THREAD

**Issue by auderson** [numba/numba#8369]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

As title.

```python
import numpy as np
from numba import njit
from numba.typed.typedlist import List
from numba.typed.typeddict import Dict


@njit
def example_list(a):
    ls = List.empty_list(a.dtype)
    ls.append(a[0])
    return ls


example_list(np.zeros(1))
```
```
- Resolution failure for literal arguments:
No implementation of function Function(<function impl_append at 0x7fc16b232af0>) found for signature:
 >>> impl_append(ListType[dtype(float64)], float64)
There are 2 candidate implementations:
    - Of which 2 did not match due to:
    Overload in function 'impl_append': File: numba/typed/listobject.py: Line 592.
      With argument(s): '(ListType[dtype(float64)], float64)':
     Rejected as the implementation raised a specific error:
       NumbaNotImplementedError: Failed in nopython mode pipeline (step: native lowering)
     Cannot cast float64 to dtype(float64): double %"arg.item"
     During: lowering "casteditem = call $2load_global.0(item, $6load_deref.2, func=$2load_global.0, args=[Var(item, listobject.py:600), Var($6load_deref.2, listobject.py:600)], kws=(), vararg=None, varkwarg=None, target=None)" at /home/auderson/miniconda3/lib/python3.8/site-packages/numba/typed/listobject.py (600)
  raised from /home/auderson/miniconda3/lib/python3.8/site-packages/numba/core/base.py:704
- Resolution failure for non-literal arguments:
None
During: resolving callee type: BoundFunction((<class 'numba.core.types.containers.ListType'>, 'append') for ListType[dtype(float64)])
During: typing of call at <ipython-input-12-6baf440651d6> (10)
File "<ipython-input-12-6baf440651d6>", line 10:
def example_list(a):
    <source elided>
    ls = List.empty_list(a.dtype)
    ls.append(a[0])
    ^
```
```python

@njit
def example_dict(a):
    d = Dict.empty(a.dtype, a.dtype)
    d[a[0]] = a[0]
    return d


example_dict(np.zeros(1))

```
```
No implementation of function Function(<built-in function setitem>) found for signature:
 
 >>> setitem(DictType[dtype(float64),dtype(float64)]<iv=None>, float64, float64)
 
There are 16 candidate implementations:
  - Of which 14 did not match due to:
  Overload of function 'setitem': File: <numerous>: Line N/A.
    With argument(s): '(DictType[dtype(float64),dtype(float64)]<iv=None>, float64, float64)':
   No match.
  - Of which 2 did not match due to:
  Overload in function 'impl_setitem': File: numba/typed/dictobject.py: Line 695.
    With argument(s): '(DictType[dtype(float64),dtype(float64)]<iv=None>, float64, float64)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   No implementation of function Function(<built-in function hash>) found for signature:
    
    >>> hash(dtype(float64))
    
   There are 4 candidate implementations:
         - Of which 2 did not match due to:
         Overload of function 'hash': File: numba/experimental/jitclass/overloads.py: Line 0.
           With argument(s): '(dtype(float64))':
          No match.
         - Of which 2 did not match due to:
         Overload in function 'hash_overload': File: numba/cpython/hashing.py: Line 48.
           With argument(s): '(dtype(float64))':
          Rejected as the implementation raised a specific error:
            TypingError: Failed in nopython mode pipeline (step: nopython frontend)
          Unknown attribute '__hash__' of type dtype(float64)
          
          File "../miniconda3/lib/python3.8/site-packages/numba/cpython/hashing.py", line 51:
              def impl(obj):
                  return obj.__hash__()
                  ^
          
          During: typing of get attribute at /home/auderson/miniconda3/lib/python3.8/site-packages/numba/cpython/hashing.py (51)
          
          File "../miniconda3/lib/python3.8/site-packages/numba/cpython/hashing.py", line 51:
              def impl(obj):
                  return obj.__hash__()
                  ^
   
     raised from /home/auderson/miniconda3/lib/python3.8/site-packages/numba/core/typeinfer.py:1086
   
   During: resolving callee type: Function(<built-in function hash>)
   During: typing of call at /home/auderson/miniconda3/lib/python3.8/site-packages/numba/typed/dictobject.py (706)
   
   
   File "../miniconda3/lib/python3.8/site-packages/numba/typed/dictobject.py", line 706:
       def impl(d, key, value):
           <source elided>
           castedval = _cast(value, valty)
           status = _dict_insert(d, castedkey, hash(castedkey), castedval)
           ^
  raised from /home/auderson/miniconda3/lib/python3.8/site-packages/numba/core/typeinfer.py:1086
During: typing of setitem at <ipython-input-3-3aa1714a8578> (4)
File "<ipython-input-3-3aa1714a8578>", line 4:
def example_dict(a):
    <source elided>
    d = Dict.empty(a.dtype, a.dtype)
    d[a[0]] = a[0]
    ^
```

**Comment by esc** [comment]
@auderson thank you, can you use type inference as a workaround:

```
import numpy as np
from numba import njit
from numba.typed.typedlist import List
from numba.typed.typeddict import Dict


@njit
def example_list(a):
    ls = List()
    ls.append(a[0])
    return ls


print(example_list(np.zeros(1)))
```

**Comment by auderson** [comment]
@esc Thanks! This works.

**Comment by stuartarchibald** [comment]
I think this issue here that the list is declared to be of the `DType` type, but the values provided are values of the type of the dtype:
```python
from numba import njit, typed, typeof, types
import numpy as np

a = np.zeros(1)

print(typeof(a[0]))
assert typeof(a[0]) == types.float64
print(typeof(a.dtype))
assert typeof(a.dtype) == types.DType(types.float64)

@njit
def foo(a):
    ls = typed.List.empty_list(a.dtype) # a list of float64 dtypes
    ls.append(a.dtype) # append a float64 dtype
    ls.append(np.dtype(np.float64)) # and another one
    return ls

foo(a)
```
but it's also a bug in that the `typed.List` typing should have caught it before it manifest as a lowering error. If you want declare the type explicitly from the array dtype, this should work, but I think would break if run without the JIT:
```python
from numba import njit, typed
import numpy as np

a = np.zeros(1)

@njit
def foo(a):
    ls = typed.List.empty_list(a.dtype.type) # use `.type`, create a list of float64s
    ls.append(a[0])
    ls.append(1.23)
    return ls

foo(a)
```
However, as @esc noted, if you can let type inference just work it out, that tends to be a good strategy.


For the `dict` case, the issue is that there's no implementation of `__hash__` registered for `dtype`. However, the sample code is also trying to put `float64` types into the dictionary values which are declared as a `types.DType(types.float64)` i.e. a `dtype` type. Something like this could make it work:
```python
from numba import njit, typed, typeof, types
import numpy as np
from numba.extending import overload_method

@overload_method(types.DType, "__hash__")
def ol_dt_hash(inst,):
    def impl(inst):
        return 1 # Mock hash value, the real hash alg would need replicating
    return impl


@njit
def foo(a):
    d = typed.Dict.empty(a.dtype, a.dtype)
    d[a.dtype] = a.dtype # values are dtypes!


a = np.zeros(1)
foo(a)
```

**Comment by auderson** [comment]
@stuartarchibald  Many thanks for this detailed reply. Actually I've tried `ls = typed.List.empty_list(a.dtype.dtype)` becasue I found `DType` has a `dtype` property😂.

**Comment by stuartarchibald** [comment]
> @stuartarchibald Many thanks for this detailed reply. Actually I've tried `ls = typed.List.empty_list(a.dtype.dtype)` becasue I found `DType` has a `dtype` propertyjoy.

No problem. Glad you've found something that works. Am going to close this as resolved and have opened #8372 to track the issue with `typed.List.append` not validating types correctly.
